### ISSUE TITLE
Scope/API Design Discussion

### ISSUE THREAD

**Issue by SkaterDad** [hyperapp/server#8]
I'd like to get some discussion going around the scope of this project, and what kind of API people are hoping for.  Feel free to suggest other questions/topics that will help shape this project :smile: 

### Scope / API
- **Scope:**
  - How much will this package do for users?
  - What features should be supported?
- **API:**
  - How is this actually used in their server code?
  - What can be configured?

### Support
- **What version(s) of Node should be officially supported?**
  - Will need the test suite to run against each version
- **Official packages with server framework integrations?**
  - Express/Hapi/Koa, etc...

**Comment by SkaterDad** [comment]
## My Take
tldr:  Make it easy to use, and allow for waiting on async actions automatically.  Node 8+

### Scope / API
Every app may have different requirements, and we probably all have different opinions, so this should be an interesting point to discuss.

I feel like most of the existing frameworks (like React/Vue) make SSR kind of difficult, especially if you have asynchronous data fetching to do on init.  Partially this is because there are infinite ways of doing everything.   "Do everything" frameworks like `Next.js` and `Nuxt.js` are born from that difficulty, but they can make debugging harder.  I think the simplicity and structure (state management/event system) of `hyperapp` makes this much easier. :smile: 

- **API (High Level)**
  - Server-specific `app(appConfig, serverConfig)` function, called once at server startup.  Performs the same initialization logic that `hyperapp` does on the client (merging mixins, binding functions, etc), but without running any events/actions.  Could also do this per request depending on perf.
    - 1st param: same config object as client version
    - 2nd param: server-specific config
    - Returns functions that will return HTML as string or stream (includes serialized state, and automatically adding "data-ssr" attribute to the root element).
    - Can also return low-level stuff like `{ state, actions, view, emit }` for people who want to do their own thing.
- **Must Have Features/Config (for my use cases)**
   - User can specify which events get emitted before rendering (default: `["init", "loaded"]`)
   - Optionally wait for async actions (this is why I think you need a custom `app()` on server)
   - Accept a template object/function, so that the serialized state & view can be put in appropriate spots in the final HTML which gets sent back.
  - Escape potentially dangerous strings (like in text nodes)

**Expand this to see usage examples:**
<details>

```js
// Mock globals as needed.
global.document = {
  getElementById: id => id
}

// Express
const server = require("express")()

// Load hyperapp config object (same as client app will use)
const appConfig = require("./appConfig")

// Define hyperapp/server configuration (defaults shown)
const serverConfig = {
  async: true,
  events: ["init", "loaded"]
}

// Import & initialize server-side app() function
const hyperapp = require("../src/app")(appConfig, serverConfig)

// Define route handlers
// Using async/await for readability

// This route renders to a string, then sends it to the client
// Template could be a simple function which accepts state & html strings
server.get("string", async (req, res) => {
  const html = await hyperapp.renderToString(({state, html}) => `
    <!doctype html>
    <html>
      <head>
        <title>SSR App</title>
        <script id="ssr-state" type="application/json">
          ${state}
        </script>
      </head>
      <body>
        <div id="app">${html}</div>
        <script src="dist/client.js"></script>
      </body>
    </html>
    `)

    res.send(html)
})

// This route pipes a stream back to the client.  This can send initial chunks sooner than the other option, and sends more as it is processed.  Template needs to be defined pieces for that to work.
server.get("stream", async (req, res) => {
  const template = {
    //Before serialized state
    head: `
      <!doctype html>
      <html>
      <head>
        <title>SSR App</title>
        <!-- css, whatever else -->`,
    //Between state & view
    neck: `
      </head>
      <body>
        <div id="app">`,
    //After view
    tail: `
        </div>
        <script src="dist/client.js"></script>
      </body>
    </html>`,
  }
  
  const stream = hyperapp.renderToStream(template)

  stream.pipe(res)
})

const port = process.env.PORT || 3000
const host = process.env.IP || "localhost"

server.listen(port, host, () =>
  console.log(`Listening at http://${host}:${port}`)
)
```
</details>

### Support
- **What version(s) of Node should be officially supported?**
  - Node 8+ has async/await and most of the newest ES features by default, so our code could be clean and shiny :sparkles: 
  - We could provide a transpiled version of the code for people on older Node versions (ex: `require('hyperapp-server/legacy')`).  With CI running tests against the older versions, this should be ok?
- **Official packages with server framework integrations?**
  - Basing everything on Node core's http request/response objects could make this package framework agnostic, and trivial to integrate with an Express/Hapi/Koa project.
  - Providing an official middleware/plugin seems like extra headaches :face_with_head_bandage:

**Comment by andyrj** [comment]
Other than the initialized once and re-use of app mentioned above, which I think would be an anti-pattern if the user doesn't build some kind of state reset.  Couldn't the rest of what you want be achieved by the users themselves, without needing a custom server side app()?

I feel like each person's app will have different needs and trying to design to anyone of them will cause issues for some other set of use cases.  But if we have the building blocks and an example mixin they can customize it to suit their purpose.  

As an example what you describe above about having some specific trigger after completing certain events fits what I am trying to say.  

Instead of that why not have the users set a value in state when their route has finished rendering/async fetching and they can test for that state in a render event handler.  I think that leaves more open in cases where users go with one of the competing component patterns that may be nested or have it's own private state or async data fetching.  Otherwise I think we are going to end up adding a lot of complexity.

**Comment by SkaterDad** [comment]
@andyrj 
> Other than the initialized once and re-use of app mentioned above, which I think would be an anti-pattern if the user doesn't build some kind of state reset.

That is a good point.  I had not been inspecting my HTML output closely enough after doing multiple requests.  The internal state object just gets mutated, so that's no good!   We can put the state reset into the library, though, so the users don't need to think about it.

> Instead of that why not have the users set a value in state when their route has finished rendering/async fetching and they can test for that state in a render event handler. I think that leaves more open in cases where users go with one of the competing component patterns that may be nested or have it's own private state or async data fetching. Otherwise I think we are going to end up adding a lot of complexity.

Running hyperapp untouched w/ a SSR mixin is still an idea that I like quite a lot, since there are big upsides.  My custom server-side `app()` function is a more opinionated solution, I suppose, so it wouldn't necessarily be useful for everyone.

Putting something in `state` when the async stuff is completed could work, so long as you're diligent about testing all of your code paths.  I still wish there was a way to get this functionality for free, though.  In my custom server app(), making `emit` an async function that `await`ed all of the action calls solved this handily.

One concern is the performance loss compared to a custom server-side `app()`.  As an example, on the server, you don't really need to run the the whole v-dom generation & patching until the final state has been achieved (assuming users aren't relying on some element `oncreate` handlers to fire actions).

Maybe some of that perf hit could be avoided by having the SSR mixin return a dummy v-node object from the "render" event until the special "ssr is ready" state is set?  Fewer function calls, and nothing to diff in the patch function.   I haven't followed the "components" discussion closely, so am not sure if this would break those?

```js
//SSR mixin
{
  events: {
      render: (state, actions, view) => {
        return state.ssr
          ? view
          : (state, actions) => ({ tag: 'h1', data: null, children: ["Not done yet!"]})
      }
  }
}
```

**Comment by andyrj** [comment]
I would like to hear what @jbucaran thinks on this.  But yes your above mixin snippet is more along the lines of what I was thinking.

@SkaterDad
I would love to see a gist of what you are thinking for the server side app().  Trying to avoid the patch passes server side sort of feels like a pre-mature optimization that will cause maintenance headaches but maybe you have away to avoid what I am worrying about.

I don't mean for any of what I am saying to be taken as a negative I just want us to end up with an awesome ssr solution.

**Comment by SkaterDad** [comment]
No negativity was assumed!  It's hard to land on a solid solution to anything without some debate & critique.   I tend to change my mind a few times before I even hit "Comment", since writing everything out makes me think critically of my own ideas, too.

I'll keep iterating on my solution before pushing to my fork of this repo.  Time permitting, I'll aim for tomorrow.

What kind of maintenance headaches are you predicting?

**Comment by andyrj** [comment]
Well I am just nervous that the more hyperapp clientside api we replicate in a server side version, the more difficult keeping feature parity will become.  If we can manage a solution that just leverages the client side app(), then we just inherit new features without having to replicate them for the server side app().  I don't have a specific example atm, but if mixins where to change, or maybe the events.  I could see that being some work to maintain.  That all also depends on how the server side app() is setup though, so I would need to see more of what you were thinking to give a more concrete example.

**Comment by SkaterDad** [comment]
@andyrj I don't my code will calm your nerves at all, but I've [pushed it to my fork](https://github.com/SkaterDad/server/tree/server-app
).  I can certainly understand how the approach I took would require more maintenance long-term.  However, I think it gives you some cool optimization opportunities since the sever code can diverge from the client lib (premature to pursue at this point, of course, without some real world experience).

The code is commented pretty thoroughly, so hopefully it's easy to follow.   The README has instructions on how to try it out with a very basic example.

**Comment by corysimmons** [comment]
It'd be nice if Hyperapp just SSR'd by default and was rebuilt from the ground up with that goal in mind.

**Comment by jbucaran** [comment]
Hi @corysimmons, how would that work? 🤔

**Comment by corysimmons** [comment]
Same way Next/Nuxt/Gatsby work I suppose. Bundle router and SSR into Hyperapp.

Might be better as a standalone repo, but my use case is almost always wanting to reach for one of those at the beginning of a project rather than almost always have to fight for fast initial page loads and not having to worry about crawlers at all.

Just a lazy suggestion. Hyperapp looks cool because it's batteries included (state management included for instance). 1 path forward seems really nice. Starting with SSR out of the box seems really popular right now so I don't think I'm alone.

**Comment by jbucaran** [comment]
@corysimmons The router is a different story, but I am not sure about SSR. We current make it easy to hydrate statically rendered HTML, which is close, and in some cases good enough SSR.

Now, since SSR only makes sense when using hyperapp on the server, adding the stuff you suggest into core wouldn't affect CDN bundle size 🎉 , and **if you are not** using SSR when building your app, dead code elimination via webpack or rollup would completely remove the SSR code anyway, so you win again.

I think we should re-purpose, archive or just delete this repo and move the SSR stuff into hyperapp. But we need a plan first.

/cc @SkaterDad

**Comment by corysimmons** [comment]
Thank you. My team is waiting on 1.0.0 then we're going to discuss migrating and building greenfield frontends with Hyperapp. SSR would be a huge argument _for_ the move.

>But we need a plan first.

I agree. I can do a little research soon and give some advice on what similar frameworks are doing. I suspect implementation won't be as difficult if we learn from their past mistakes.

**Comment by jbucaran** [comment]
> I agree. I can do a little research soon and give some advice on what similar frameworks are doing. I suspect implementation won't be as difficult if we learn from their past mistakes.

Please do! 🙏

**Comment by JorgeBucaran** [comment]
Hi @SkaterDad. I have no use for this package, and it's been a while since we had any significant updates. Would you like me to transfer it to you or do you have no use for it yourself as well?

**Comment by SkaterDad** [comment]
@JorgeBucaran I'm not in need of it either.   If/when I need SSR, [hyperapp-render](https://github.com/frenzzy/hyperapp-render)  looks like a great option.

**Comment by corysimmons** [comment]
Maybe we can just make a CLI like vue-cli that will create a Hyperapp app. Add flag for something like `--ssr`.

**Comment by JorgeBucaran** [comment]
@corysimmons Haha dammit 🙃
